### PR TITLE
[X11] Create windows with a parent window

### PR DESCRIPTION
**Pull request by t-sin**
This PR can create windows with a user-specified parent window. It's intended to realize what discussed in issue https://github.com/rust-windowing/winit/issues/159 for X11.

This is a retry https://github.com/rust-windowing/winit/pull/2096. In that time I was not aware CI fails so passed long time, the master branch proceeds so much from that PR so I cannot follow changes.

## TODO

- [x] Tested on all platforms changed
    - [x] X11
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

### CODE REVIEW

File: examples/child_window.rs
Comment (t-sin) [inline examples/child_window.rs:20]: Need some help.

 In `examples/child_window.rs`, platform-specific functions configured to build only when `x11` feature but fails other feature like windows like this: https://github.com/rust-windowing/winit/runs/5859985935?check_suite_focus=true#step:9:129 but I don't know why...

**Comment by t-sin** [comment]
All checks have passed! It's truly ready for review :confetti_ball:

File: examples/child_window.rs
Comment (dhardy) [inline examples/child_window.rs:23]: Is this position relative to the parent's inner coordinate system or the screen or something else?

Comment (t-sin) [reply examples/child_window.rs:23]: The position is relative to parent's inner one.

Comment (t-sin) [reply examples/child_window.rs:23]: Updating docs that is very nice idea. I will update it please wait.

Comment (t-sin) [reply examples/child_window.rs:23]: How about this https://github.com/rust-windowing/winit/pull/2246/commits/daf680c87aec2d4f666c4fa119a206a9cc213a31 ?

**Comment by t-sin** [comment]
This PR followed the HEAD of master. Including document updates.

**Comment by t-sin** [comment]
What is missing in this PR?

**Comment by t-sin** [comment]
What does this PR stop? I’ll try to resolve that if I can.

**Comment by madsmtm** [comment]
We don't have a maintainer for X11 atm, so there's not really anyone to neither review nor accept the PR, sadly

**Comment by t-sin** [comment]
Oh... It make sense.

**Review by dhardy — COMMENTED** [review]
Not from your PR, but maybe you could also fix these errors?

> `#[doc(hidden)]` is ignored on trait impl items

The code here... really, there's not much to review, besides the example.

Being able to visualise things is always nice, so I added [softbuffer](https://crates.io/crates/softbuffer) as a dev-dependency and modified the example. See my commits here: https://github.com/dhardy/winit/commits/with-parent-window-in-x11

x11-parent-window

File: src/platform/unix.rs
Comment (dhardy) [inline src/platform/unix.rs:279]: One other point worth making: is it possible to use the `winit::WindowId` here?

**Comment by t-sin** [comment]
Sorry I did my project... I can do this PR starting this week end maybe.

**Comment by madsmtm** [comment]
Re adding `softbuffer` to our examples: I think it makes more sense to direct people toward it via. documentation, mostly since they don't support the same platforms as we do (`glutin` would be a better fit here), and is not a crate we're really in control of.

**Comment by kchibisov** [comment]
>Re adding softbuffer to our examples: I think it makes more sense to direct people toward it via. documentation, mostly since they don't support the same platforms as we do (glutin would be a better fit here), and is not a crate we're really in control of.

Using softbuffer should resolve issues with examples on Wayland, so it should simplify testing.

**Comment by dhardy** [comment]
> mostly since they don't support the same platforms as we do (glutin would be a better fit here)

Will there ever be a perfect fit? Moreover, I'm skeptical that using a complex hardware-accelerated API as the default for visualizations is the most reliable choice, even if GL does work basically everywhere.

Given some direction I'm happy to author a different PR regarding softbuffer-in-winit, but it's simple enough that anyone here could do this.

**Comment by kchibisov** [comment]
>Given some direction I'm happy to author a different PR regarding softbuffer-in-winit, but it's simple enough that anyone here could do this.

If you feel like it you can go for it for sure.

**Comment by dhardy** [comment]
Quite aside from the use of softbuffer (which is now #2353), what is the status of this PR?

- Winit has no X11 maintainer; nevertheless there isn't much code here and I left some review comments (still unaddressed)
- Does Winit have any policies on additional platform-specific APIs?
- From what I understand, `WindowId` on X11 has type [`x11-dl::xlib::Window`](https://docs.rs/x11-dl/latest/x11_dl/xlib/type.Window.html) which is the same value returned by `fn xlib_window`, thus it may make sense to pass a `winit::WindowId` into `with_x11_parent` (or it may not: `WindowId` is an enum over X11 and Wayland identifiers)

File: src/platform/unix.rs
Comment (maroider) [reply src/platform/unix.rs:279]: `usize` is likely the wrong type to use here. X11 window IDs are `u32`s, as far as I can tell (despite what `x11-dl`'s typedefs would have you believe).

As of #2351, we can make a `WindowId` from arbitrary `u64`s, so using it instead should be fine.

**Comment by maroider** [comment]
> what is the status of this PR?

Aside from the lack of an X11 maintainer, I suppose we're waiting on @t-sin to respond to your review.

> Winit has no X11 maintainer

Yep, this does suck, and it's a position I may have to fill myself going forward, unless a more interested (and *actually qualified*) candidate appears.

> Does Winit have any policies on additional platform-specific APIs?

IIRC, it's mostly just that such APIs are best suited for things we cannot create a meaningful cross-platform abstraction over. I don't think there's anything written down about APIs we'd like to have a cross-platform abstraction over, but currently haven't invested the time in creating.

File: CHANGELOG.md
Comment (t-sin) [reply CHANGELOG.md]: Oh sorry. Removed!

File: src/platform/unix.rs
Comment (t-sin) [reply src/platform/unix.rs:299]: > One other point worth making: is it possible to use the winit::WindowId here?

> As of https://github.com/rust-windowing/winit/pull/2351, we can make a WindowId from arbitrary u64s, so using it instead should be fine.

Good idea, thanks! I'll do that.

**Comment by t-sin** [comment]
Status reporting. I'm working visualized child window creations and found a bug https://github.com/rust-windowing/winit/pull/2363. This PR depends that PR so I'm waiting some reviews.

**Comment by t-sin** [comment]
Visualizing child window is finished. I've done for all review points, so this PR is ready for review.

**Comment by t-sin** [comment]
Now CI fails in like i386-linux. I'm not familiar with X11, [`XID` sometimes typedefs as a pointer](https://github.com/freedesktop/xorg-xserver/search?q=typedef+xid)...? CI fails in i386-linux because [`XID` is not u32](https://github.com/rust-windowing/winit/runs/7205283569?check_suite_focus=true).

To solve it roughly, it need implementing `From<WindowId> for u32` I think but any idea?

**Comment by kchibisov** [comment]
>To solve it roughly, it needs implementing From<WindowId> for u32 or for usize I think, but any idea?

(id as u64).into().

**Comment by t-sin** [comment]
Now CI on Linux is passed. I don't know why Windows CI fail.

**Comment by maroider** [comment]
It looks like it's all `softbuffer`'s fault.

**Comment by dhardy** [comment]
Of course. I had hoped `softbuffer` would be a simple and reliable dependency, but clearly it is not.

**Comment by t-sin** [comment]
All CI have passed.

File: src/platform_impl/linux/x11/window.rs
Comment (notgull) [inline src/platform_impl/linux/x11/window.rs:126]: Isn't a window guaranteed to be a 32 bit value? This pattern probably isn't exclusive to this PR, but I don't think the `try_*` items are necessary, since `as` will lead to no loss of precision.

File: CHANGELOG.md
Comment (t-sin) [inline CHANGELOG.md:75]: Ooh sorry.

File: src/platform_impl/linux/x11/window.rs
Comment (t-sin) [reply src/platform_impl/linux/x11/window.rs:126]: When I read x11 code little, I found that the window IDs sometimes are typedefed as a pointer (https://github.com/rust-windowing/winit/pull/2246#issuecomment-1175601618) so I don't think the window IDs is not guaranteed to be a 32-bit value.

Comment (notgull) [reply src/platform_impl/linux/x11/window.rs:126]: That's strange, because `libxcb` relies on XIDs being 32-bit, so any change that would make an XID longer than 32 bits would break `libxcb`. In addition, the X11 protocol [defines XIDs to be 32 bits](https://www.x.org/docs/XProtocol/proto.pdf) (see page 5).

File: CHANGELOG.md
Comment (t-sin) [reply CHANGELOG.md]: Fixed.

File: src/platform_impl/linux/x11/window.rs
Comment (t-sin) [reply src/platform_impl/linux/x11/window.rs:126]: Thanks, I made a misunderstanding for it. So these `try_into()` are not neccesary for me too, I'll fix it.

Comment (t-sin) [reply src/platform_impl/linux/x11/window.rs:126]: Fix that!

Comment (maroider) [reply src/platform_impl/linux/x11/window.rs:126]: > `libxcb` relies on XIDs being 32-bit, so any change that would make an XID longer than 32 bits would break `libxcb`

Agreed, however, `x11-dl` has had a seemingly broken pointer-width-reliant type alias for a while now (making their `XID=c_ulong`, which is 64-bit on 64-bit Linux) yet it somehow works for us.

 > Fix that!

I'm a bit confused now. Do you mean you've fixed it, or that you're going to fix it? And if you mean that you've fixed it, then I'm a bit confused as to why you're using `u64`s and `cfg(target_pointer_width = "64")` still.

Comment (t-sin) [reply src/platform_impl/linux/x11/window.rs:126]: > I'm a bit confused now. 

Sorry for my rough English. I want to say "Fixed". And I agree your comment: there's no reason to use `cfg(...)`.

File: CHANGELOG.md
Comment (t-sin) [reply CHANGELOG.md]: Thanks! Fixed it.

File: src/platform_impl/linux/x11/window.rs
Comment (t-sin) [reply src/platform_impl/linux/x11/window.rs:126]: I think it's fixed.

Comment (t-sin) [reply src/platform_impl/linux/x11/window.rs]: Right, before this fix https://github.com/rust-windowing/winit/pull/2246/commits/62c9408f28782040d5de30b224bfdafbd543f795 that compiler conditional remains. Now that's removed so I think it's good.

**Comment by t-sin** [comment]
I stop resolving conflicts in this PR. Notify me when it can be review for merge.

**Review by kchibisov — CHANGES_REQUESTED** [review]
Could you rebase?

**Comment by t-sin** [comment]
Rebasing has done!

File: CHANGELOG.md
Comment (madsmtm) [inline CHANGELOG.md:21]: Nit:
```suggestion
- On X11, added `WindowExtX11::with_x11_parent` to create child windows.
```

On that note, maybe rename to just `with_parent`?

File: src/platform/x11.rs
Comment (madsmtm) [inline src/platform/x11.rs:4]: You can remove these now

File: CHANGELOG.md
Comment (t-sin) [reply CHANGELOG.md]: Thanks! Fixed and Renamed it.

File: src/platform/x11.rs
Comment (t-sin) [reply src/platform/x11.rs:4]: True that. Fixed.

Comment (madsmtm) [reply src/platform/x11.rs:4]: This... Does not seem fixed to me, could you take a second look?

Comment (t-sin) [reply src/platform/x11.rs]: Sorry... I saw a different line... Now removed that `#cfg[...]`

File: src/platform/x11.rs
Comment (madsmtm) [inline src/platform/x11.rs:175]: Nit: Also remove this

Comment (t-sin) [reply src/platform/x11.rs]: Fixed!!

**Review by kchibisov — CHANGES_REQUESTED** [review]
Could you rebase. Should be good after that.

**Comment by t-sin** [comment]
Rebased!

File: src/platform/x11.rs
Comment (kchibisov) [reply src/platform/x11.rs:175]: This is still not removed.

File: src/platform/x11.rs
Comment (kchibisov) [inline src/platform/x11.rs:176]: We're inside the X11 extension.

```suggestion
    /// Build window with parent window.
```

File: src/platform_impl/linux/x11/window.rs
Comment (kchibisov) [inline src/platform_impl/linux/x11/window.rs:123]: Remove this comment entirely, there's no value in it. The WindowID thing is clear.

File: src/platform_impl/linux/x11/window.rs
Comment (kchibisov) [inline src/platform_impl/linux/x11/window.rs:128]: This comment also feels a bit strange. You can say `WindowId is XID under the hood which doesn't exceed u32, so the conversion is lossless`.

**Review by kchibisov — CHANGES_REQUESTED** [review]

**Comment by t-sin** [comment]
Thanks @kchibisov for reviewing my some nits. I think these are fixed now (checked all diffs in the "Files changed" tab), but `cargo check` fails in my machine so I wait CI results...

File: src/platform_impl/linux/x11/window.rs
Comment (t-sin) [reply src/platform_impl/linux/x11/window.rs]: fixed

File: src/platform_impl/linux/x11/window.rs
Comment (t-sin) [reply src/platform_impl/linux/x11/window.rs]: fixed

File: src/platform/x11.rs
Comment (t-sin) [reply src/platform/x11.rs]: thanks! fixed.

**Review by kchibisov — APPROVED** [review]

**Pull request merged**
